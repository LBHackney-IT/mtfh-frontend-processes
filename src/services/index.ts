export { default as locale } from "./locale";
export { default as menu } from "./menu";
export * as processes from "./processes";
export * from "./config";
