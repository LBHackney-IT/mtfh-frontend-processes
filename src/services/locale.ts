import reactHtmlParser from "react-html-parser";

import { AssetAddress } from "@mtfh/common/lib/api/asset/v1";

const locale = {
  backButton: "Back",
  bookAppointment: "Book Appointment",
  next: "Next",
  cancel: "Cancel",
  cancelProcess: "Cancel process",
  change: "Change",
  views: {
    soleToJoint: {
      title: "Sole tenant requests a joint tenure",
      steps: {
        selectTenant: "Select person",
        checkEligibility: "Eligibility Checks",
        supportingDocuments: "Supporting documents",
        finish: "Finish",
        breachOfTenancy: "Breach of tenure check",
        requestDocuments: "Request Documents",
        reviewDocuments: "Review Documents",
        submitCase: "Submit case",
      },
      actions: {
        reassignCase: "Reassign case",
        cancelProcess: "Cancel process",
        caseActivityHistory: "Case activity history",
      },
    },
    selectTenants: {
      selectTenantHint:
        "This person will be asked for proof of relationship e.g. marriage or civil partnership certificate.",
      selectTenantLabel: "Who do you want to add as a joint tenant?",
      addToTenureText: "If the person you want to add is not listed you must first",
      addToTenureLink: "add them to the tenure",
      noHouseholdMembersOver18:
        "There are no household members over the age of 18 associated to this.",
    },
    processesMenu: {
      selectProcessDescription: () =>
        reactHtmlParser(
          "<p>Start a process by selecting a process category below. <br /> Where a process is not yet supported in <strong>Manage My Home</strong> these links will open the appropriate interim Google form.</p>",
        ),
    },
    checkEligibility: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Passed automatic eligibilty checks",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    breachOfTenancy: {
      autoCheckIntro:
        "Manage My Home has used tenure, person and property information to automatically check eligibility for this application.",
      passedChecks: "Eligibility checks passed",
      failedChecks: "Not eligible for a sole to joint tenure",
    },
    reviewDocuments: {
      passedChecks: "Eligibility checks passed",
      documentsRequested: "Documents requested",
      checkSupportingDocumentsAppointment:
        "I need to make an appointment to check supporting documents",
      appointmentScheduled: "Office appointment scheduled",
      seenPhotographicId:
        "I confirm that an outcome letter has been sent to the resident",
      seenSecondId:
        "I confirm I have seen a second form of ID (does not have to be photographic)",
      isNotInImmigrationControl:
        "I confirm that the prospective tenant is not subject to immigration control under the Asylum And Immigration Act 1996",
      seenProofOfRelationship:
        "I confirm I have seen proof of relationship to the existing tenant",
      incomingTenantLivingInProperty:
        "I confirm I have seen 3 separate documents proving the proposed tenant has been living at the property for a minimum of 12 months",
    },
  },
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this type.",
    unableToFindState: "Invalid State",
    unableToFindStateDescription: "We have been unable to process a state of this type.",
  },
  title: "Processes",
  loadingText: "Submitting...",
  components: {
    startProcess: {
      buttonLabel: "Start process",
      thirdPartyHeading: "Sharing information with third parties",
    },
    entitySummary: {
      tenurePaymentRef: "Tenure payment ref",
      address: (assetAddress: AssetAddress): string => {
        const {
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        } = assetAddress;

        return [
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,

          addressLine4,
          postCode,
        ]
          .filter((addressLine) => !!addressLine)
          .join(" ");
      },
    },
  },
};

export default locale;
