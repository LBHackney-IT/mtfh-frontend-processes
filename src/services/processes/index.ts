export * from "./sole-to-joint";
export * from "./change-of-name";
