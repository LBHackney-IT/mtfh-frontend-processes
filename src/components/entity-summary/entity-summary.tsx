import { useEffect } from "react";
import { Link as RouterLink } from "react-router-dom";

import { locale } from "../../services";

import { useAsset } from "@mtfh/common/lib/api/asset/v1";
import { Asset } from "@mtfh/common/lib/api/asset/v1/types";
import { usePerson } from "@mtfh/common/lib/api/person/v1";
import { RelatedEntity } from "@mtfh/common/lib/api/process/v1";
import { useTenure } from "@mtfh/common/lib/api/tenure/v1";
import {
  Center,
  ErrorSummary,
  Heading,
  Link,
  Spinner,
} from "@mtfh/common/lib/components";

import "./styles.scss";

const { components } = locale;

interface ComponentProps {
  id: string;
  // eslint-disable-next-line react/no-unused-prop-types
  config?: Record<string, any>;
  // eslint-disable-next-line react/no-unused-prop-types
  setRelatedEntities?: (relatedEntities: RelatedEntity[]) => void;
  // eslint-disable-next-line react/no-unused-prop-types
  setAsset?: (asset: Asset) => void;
}

const TenureSummary = ({
  id,
  config = {},
  setRelatedEntities,
  setAsset,
}: ComponentProps) => {
  const { incomingTenant } = config;
  const { error, data: tenure } = useTenure(id);
  const { data: asset } = useAsset(tenure?.tenuredAsset?.id || null);

  useEffect(() => {
    if (setAsset && asset) {
      setAsset(asset);
    }
  }, [asset, setAsset]);

  const tenant = tenure?.householdMembers?.find((m) => m.isResponsible);

  useEffect(() => {
    if (setRelatedEntities) {
      const relatedEntities: RelatedEntity[] = [];

      if (tenant) {
        relatedEntities.push({
          id: tenant.id,
          targetType: "person",
          subType: "tenant",
          description: tenant.fullName,
        });
      }

      if (tenure) {
        relatedEntities.push({
          id: tenure.id,
          targetType: "tenure",
          description: tenure.paymentReference,
        });
        if (tenure.tenuredAsset) {
          relatedEntities.push({
            id: tenure.tenuredAsset.id,
            targetType: "asset",
            description: tenure.tenuredAsset.fullAddress,
          });
        }
      }

      if (relatedEntities.length > 0) {
        setRelatedEntities(relatedEntities);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [tenant]);

  if (error) {
    return (
      <ErrorSummary
        id="entity-summary"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  if (!tenure) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <Heading className="entity-summary__tenure-heading" variant="h2">
      <div>
        {components.entitySummary.tenurePaymentRef} {tenure?.paymentReference}
      </div>
      <div>{tenure?.tenuredAsset?.fullAddress}</div>
      {tenant && (
        <div>
          <Link
            as={RouterLink}
            to={`/person/${tenant.id}`}
            variant="link"
            target="_blank"
          >
            {tenant?.fullName}
          </Link>
          {incomingTenant && ` adding ${incomingTenant.fullName}`}
        </div>
      )}
    </Heading>
  );
};

const AssetsSummary = ({ id, setAsset }: ComponentProps) => {
  const { error, data: asset } = useAsset(id);

  if (error) {
    return (
      <ErrorSummary
        id="entity-summary"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  if (!asset) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  if (setAsset) {
    setAsset(asset);
  }

  return (
    <h2 className="lbh-heading-h2">
      {components.entitySummary.address(asset.assetAddress)}
    </h2>
  );
};

const PersonSummary = ({ id, setRelatedEntities, setAsset }: ComponentProps) => {
  const { error, data: person } = usePerson(id);

  const fullName = person ? `${person.firstName} ${person.surname}` : "";
  const activeTenure = person?.tenures?.find((tenure) => tenure.isActive);

  const { data: asset } = useAsset(activeTenure?.assetId || null);

  useEffect(() => {
    if (setAsset && asset) {
      setAsset(asset);
    }
  }, [asset, setAsset]);

  useEffect(() => {
    if (setRelatedEntities) {
      const relatedEntities: RelatedEntity[] = [];

      if (person) {
        if (activeTenure) {
          relatedEntities.push({
            id: person.id,
            targetType: "person",
            subType: "tenant",
            description: fullName,
          });
          relatedEntities.push({
            id: activeTenure.id,
            targetType: "tenure",
          });
          relatedEntities.push({
            id: activeTenure.assetId,
            targetType: "asset",
            description: activeTenure.assetFullAddress,
          });
        }
      }

      if (relatedEntities.length > 0) {
        setRelatedEntities(relatedEntities);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [person]);

  if (error) {
    return (
      <ErrorSummary
        id="entity-summary"
        title={locale.errors.unableToFetchRecord}
        description={locale.errors.unableToFetchRecordDescription}
      />
    );
  }

  if (!person) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <h2 className="lbh-heading-h2">
      <div>
        <Link as={RouterLink} to={`/person/${person.id}`} variant="link" target="_blank">
          {fullName}
        </Link>
      </div>
    </h2>
  );
};

interface EntitySummaryProps {
  id: string;
  // eslint-disable-next-line react/no-unused-prop-types
  config?: Record<string, any>;
  type: "tenure" | "property" | "person";
  setRelatedEntities?: (relatedEntities: RelatedEntity[]) => void;
  setAsset?: (asset: Asset) => void;
}

export const EntitySummary = ({
  id,
  type,
  config,
  setRelatedEntities,
  setAsset,
}: EntitySummaryProps) => {
  const Components = {
    tenure: TenureSummary,
    property: AssetsSummary,
    person: PersonSummary,
  };
  const Component = Components[type];

  if (!Component) return null;

  return (
    <Component
      id={id}
      config={config}
      setRelatedEntities={setRelatedEntities}
      setAsset={setAsset}
    />
  );
};
