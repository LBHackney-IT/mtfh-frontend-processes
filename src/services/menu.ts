import * as processes from "./processes";

import { Asset } from "@mtfh/common/lib/api/asset/v1/types";
import { Tenure } from "@mtfh/common/lib/api/tenure/v1/types";

interface MenuProps {
  label: string;
  link: string;
  processes?: { label: string; link: string; targetType: string }[];
  getPplQuery?: (context: any) => string;
}

const getProcessLink = (processName: string) => {
  const process = processes[processName];
  return {
    targetType: process.targetType,
    label: process.title,
    link: `/processes/${process.processName}/start`,
  };
};

const getPplQueryForRequestALegalReferral = ({ tenure, asset }): string => {
  if (!tenure || !asset) {
    return "";
  }
  const {
    householdMembers,
    tenuredAsset: { propertyReference },
  } = tenure as Tenure;
  const {
    assetAddress: { addressLine1, postCode },
  } = asset as Asset;
  const tenantNames = householdMembers
    .filter((member) => member.isResponsible)
    .map((member) => member.fullName)
    .join(", ");
  const address = addressLine1;
  const query = `usp=pp_url&entry.1699199605=${tenantNames}&entry.2086190845=${address}&entry.1956036523=${postCode}&entry.1140527432=${propertyReference}`;
  const encodedQuery = encodeURI(query);
  return encodedQuery;
};

const getPplQueryTenancyChange = ({ tenure }): string => {
  if (!tenure) {
    return "";
  }
  const {
    householdMembers,
    paymentReference,
    tenuredAsset: { fullAddress, propertyReference },
  } = tenure as Tenure;

  const tenant = householdMembers.find((member) => member.isResponsible);
  const query = `usp=pp_url&entry.1942863541=${tenant?.fullName}&entry.1090481808=${fullAddress}&entry.321634337=${propertyReference}&entry.582205322=${paymentReference}`;
  const encodedQuery = encodeURI(query);
  return encodedQuery;
};

const menu: MenuProps[] = [
  {
    label: "Sole to Joint",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    processes: [
      getProcessLink(processes.soletojoint.processName),
      getProcessLink(processes.changeOfName.processName),
    ],
  },
  {
    label: "Other tenancy changes",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    getPplQuery: getPplQueryTenancyChange,
  },
  {
    label: "Record a tenancy / audit homecheck",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfyXscVvgCBhbBBe7ADpV6cUqJoKrrTbU0YYRB_uXS2J4AHMQ/viewform",
  },
  {
    label: "Find legal referral response",
    link: "https://docs.google.com/spreadsheets/d/1fd6FwBXvBo2DmX8qZaXUr7hX9sE6Zr7HZsPlA_OO8Hs/edit#gid=1251146140",
  },
  {
    label: "Request a legal referral",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScpHsoJQgq112HIMbW06YAFpggijwQuODJM483HGbX_PMfCMw/viewform",
    getPplQuery: getPplQueryForRequestALegalReferral,
  },
  {
    label: "Record notice served",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfRSMia4FeT6WPjrmRlsLpOXyJ4tFs3Sv_O0toxUbv_FmIjKw/viewform",
  },
  {
    label: "Rehousing",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScPXSs4-xxTH2N5dUwtfmEl5mpu7kxZ6mHe86Qy_rlDNu-GbA/viewform",
  },
  {
    label: "Person",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScxrWFLF7M7A3B1yzY-LZFXkV0P0qv6igcQ4wEgWHQk8DaUCw/viewform",
  },
  {
    label: "Report Japanese knotweed",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSceU2H-VWi2ewaEiOAkOWULVFU2DoPSen8SRPyJrmQtRpEhfQ/viewform?gxids=7628",
  },
  {
    label: "Report overgrown tree",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfLmlLraDyOfqYhdjkOfkNWWlHUUep05G5n0GN-ewizN0Fbuw/viewform?gxids=7628",
  },
  {
    label: "Report overgrown garden",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSc3mOUUEQhdRiDsS0k9XDWlfGJG0CEFAC2rnC1sYmM-idt-Nw/viewform?gxids=7628",
  },
];

export default menu;
