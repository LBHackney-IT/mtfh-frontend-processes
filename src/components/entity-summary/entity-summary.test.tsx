import {
  getAssetV1,
  getPersonV1,
  getTenureV1,
  mockActiveTenureV1,
  mockAssetV1,
  mockPersonV1,
  render,
  server,
} from "@hackney/mtfh-test-utils";
import { screen } from "@testing-library/react";

import { locale } from "../../services";
import { EntitySummary } from "./entity-summary";

describe("EntitySummary", () => {
  describe("for tenure", () => {
    test("it renders tenure summary details", async () => {
      const { container } = render(
        <EntitySummary id={mockActiveTenureV1.id} type="tenure" />,
        {
          url: `/processes/soletojoint/start/tenure/${mockActiveTenureV1.id}`,
          path: "/processes/:processName/start/:targetType/:targetId",
        },
      );

      await expect(
        screen.findByText(locale.components.entitySummary.tenurePaymentRef, {
          exact: false,
        }),
      ).resolves.toBeInTheDocument();

      await expect(
        screen.findByText(mockActiveTenureV1.paymentReference, { exact: false }),
      ).resolves.toBeInTheDocument();

      await expect(
        screen.findByText(mockActiveTenureV1?.tenuredAsset?.fullAddress, {
          exact: false,
        }),
      ).resolves.toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });

    test("it renders tenure summary details with an incoming tenant", async () => {
      const incomingTenant = {
        fullName: "IncomingTenant Test",
      };
      render(
        <EntitySummary
          id={mockActiveTenureV1.id}
          type="tenure"
          config={{ incomingTenant }}
        />,
        {
          url: `/processes/soletojoint/start/tenure/${mockActiveTenureV1.id}`,
          path: "/processes/:processName/start/:targetType/:targetId",
        },
      );

      await expect(
        screen.findByText(locale.components.entitySummary.tenurePaymentRef, {
          exact: false,
        }),
      ).resolves.toBeInTheDocument();

      await expect(
        screen.findByText(mockActiveTenureV1.paymentReference, { exact: false }),
      ).resolves.toBeInTheDocument();

      await expect(
        screen.findByText(mockActiveTenureV1?.tenuredAsset?.fullAddress, {
          exact: false,
        }),
      ).resolves.toBeInTheDocument();

      await expect(
        screen.findByText(`adding ${incomingTenant.fullName}`),
      ).resolves.toBeInTheDocument();
    });

    test("it renders an error if tenure details can't be fetched", async () => {
      server.use(getTenureV1("error", 500));
      render(<EntitySummary id={mockActiveTenureV1.id} type="tenure" />, {
        url: `/processes/soletojoint/start/tenure/${mockActiveTenureV1.id}`,
        path: "/processes/:processName/start/:targetType/:targetId",
      });

      await expect(
        screen.findByText(locale.errors.unableToFetchRecord),
      ).resolves.toBeInTheDocument();
      await expect(
        screen.findByText(locale.errors.unableToFetchRecordDescription),
      ).resolves.toBeInTheDocument();
    });
  });

  describe("for person", () => {
    test("it renders person summary details", async () => {
      const { container } = render(<EntitySummary id={mockPersonV1.id} type="person" />, {
        url: `/processes/soletojoint/start/person/${mockPersonV1.id}`,
        path: "/processes/:processName/start/:targetType/:targetId",
      });

      await expect(
        screen.findByText(mockPersonV1?.firstName, { exact: false }),
      ).resolves.toBeInTheDocument();

      await expect(
        screen.findByText(mockPersonV1?.surname, { exact: false }),
      ).resolves.toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });

    test("it renders an error if person details can't be fetched", async () => {
      server.use(getPersonV1("error", 500));
      render(<EntitySummary id={mockPersonV1.id} type="person" />, {
        url: `/processes/soletojoint/start/person/${mockPersonV1.id}`,
        path: "/processes/:processName/start/:targetType/:targetId",
      });

      await expect(
        screen.findByText(locale.errors.unableToFetchRecord),
      ).resolves.toBeInTheDocument();
      await expect(
        screen.findByText(locale.errors.unableToFetchRecordDescription),
      ).resolves.toBeInTheDocument();
    });
  });

  describe("for asset", () => {
    test("it renders property summary details", async () => {
      const { container } = render(
        <EntitySummary id={mockAssetV1.id} type="property" />,
        {
          url: `/processes/soletojoint/start/property/${mockAssetV1.id}`,
          path: "/processes/:processName/start/:targetType/:targetId",
        },
      );
      await expect(
        screen.findByText(
          locale.components.entitySummary.address(mockAssetV1.assetAddress),
          {
            exact: false,
          },
        ),
      ).resolves.toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });

    test("it renders an error if property details can't be fetched", async () => {
      server.use(getAssetV1("error", 500));
      render(<EntitySummary id={mockAssetV1.id} type="property" />, {
        url: `/processes/soletojoint/start/property/${mockAssetV1.id}`,
        path: "/processes/:processName/start/:targetType/:targetId",
      });

      await expect(
        screen.findByText(locale.errors.unableToFetchRecord),
      ).resolves.toBeInTheDocument();
      await expect(
        screen.findByText(locale.errors.unableToFetchRecordDescription),
      ).resolves.toBeInTheDocument();
    });

    test("it does not throw if incorrect type is passed", async () => {
      expect(() => {
        // @ts-ignore
        render(<EntitySummary id={mockAssetV1.id} type="other" />, {
          url: `/processes/soletojoint/start/property/${mockAssetV1.id}`,
          path: "/processes/:processName/start/:targetType/:targetId",
        });
      }).not.toThrow();
    });
  });
});
