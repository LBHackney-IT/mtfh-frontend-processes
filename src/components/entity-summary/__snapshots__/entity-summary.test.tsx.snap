// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntitySummary for asset it renders property summary details 1`] = `
<div>
  <h2
    class="lbh-heading-h2"
  >
    1 Newcome House 151 Price Street Borders Port Leoraville Falkland Islands (Malvinas) PN NEE
  </h2>
</div>
`;

exports[`EntitySummary for person it renders person summary details 1`] = `
<div>
  <h2
    class="lbh-heading-h2"
  >
    <div>
      <a
        class="govuk-link lbh-link lbh-link--no-visited-state"
        href="/person/6fbe024f-2316-4265-a6e8-d65a837e308a"
        target="_blank"
      >
        Haylee Macejkovic
      </a>
    </div>
  </h2>
</div>
`;

exports[`EntitySummary for tenure it renders tenure summary details 1`] = `
<div>
  <h2
    class="lbh-heading-h2 entity-summary__tenure-heading"
  >
    <div>
      Tenure payment ref
       
      mt0xzvjp6g
    </div>
    <div>
      72251 Champlin Trafficway, 99647
    </div>
    <div>
      <a
        class="govuk-link lbh-link lbh-link--no-visited-state"
        href="/person/a302fe28-e5c3-4c84-a3e6-cd575e65093f"
        target="_blank"
      >
        Global Web Consultant Alverta Denesik
      </a>
    </div>
  </h2>
</div>
`;
